posthog/models/history_logging.py

```diff
@@ -0,0 +1,196 @@
+                if current_key not in previous.state:
+                    history.append(
+                        HistoryListItem(
+                            email=_safely_read_email(current),
+                            name=_safely_read_first_name(current),
+                            action=f"{history_type}_{current_key}_changed",
+                            detail={
+                                "id": current.item_id,
+                                "key": current.state["key"],
+                                "to": current.state[current_key],
+                            },
+                            created_at=current.versioned_at.isoformat(),
+                        )
+                    )
+                elif current.state[current_key] != previous.state[current_key]:
+                    history.append(
+                        HistoryListItem(
+                            email=_safely_read_email(current),
+                            name=_safely_read_first_name(current),
+                            action=f"{history_type}_{current_key}_changed",
+                            detail={
+                                "id": current.item_id,
+                                "key": current.state["key"],
+                                "from": previous.state[current_key],
+                                "to": current.state[current_key],
+                            },
+                            created_at=current.versioned_at.isoformat(),
+                        )
+                    )
+
+            for previous_key in previous.state:
+                if previous_key not in current.state:
+                    history.append(
+                        HistoryListItem(
+                            email=_safely_read_email(current),
+                            name=_safely_read_first_name(current),
+                            action=f"{history_type}_{previous_key}_deleted",
+                            detail={
+                                "id": current.item_id,
+                                "key": current.state["key"],
+                                "deleted": previous.state[previous_key],
+                            },
+                            created_at=current.versioned_at.isoformat(),
+                        )
+                    )
+        elif previous is None and current.action != "create":
+            history.append(
+                HistoryListItem(
+                    email="history.hog@posthog.com",
+                    name="history hog",
+                    action=f"{history_type}_imported",
+                    detail={"id": current.item_id, "key": current.state["key"]},
+                    created_at=current.versioned_at.isoformat(),
+                )
+            )
+
+    return history
+
+
+def _safely_read_email(version: Optional[HistoricalVersion]) -> Optional[str]:
+    if version and version.created_by:
+        return version.created_by.email
```

Comment:
Will this result in N+1 queries? Can we add a test that validates that we have O(n) queries.

Reply:
Turns out it would have done! 😱 

Added `select_related` and a test to protect regression